### DISCUSSION TITLE
Trouble with return type of generic funcion

### DISCUSSION THREAD

**Discussion by datenreisender** [colinhacks/zod#2858] [Q&A] [answered]
I read [the description how to type functions which accept a schema](https://zod.dev/?id=writing-generic-functions) but I still fail to type my case both correctly and easily.

My context: I want to write a function which accepts a schema and returns a new function which takes a string, parses that string as JSON and then with the given schema. If there is an error, that error is post-processed.

Here is my resulting code (also available [in a playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcBEyEAJvgNwBQFAxhAHYDO8MUAho5tCAKJQ5RwAvHAAUAUz7QAXHF78AlEIB8cCfwB0IMQwasA5mPUwIAVTBgJAYVYMxI+ZRr0mcBrjEBlagAsxIVkIo6hAARgBWYtQwIggUcHAArrZQdKxaMsjqTFDAdHr2ADQUaA5UTozwYKxQtgDqwDDeAApQYjAwwJjAYsRy0ACy2roGQnFwADwAKqoAHjBidMQMQQBaJJMAnhYAgnQbSiIMPn6sMpOKgkpjIrR083cy2bl6Fyqx8fG3Lq0MCQA28GER18-iyrEwYia1VsIgAUh4APIAOXUVRqdlu9xg8lKY3inVEP3+MCyCWo1CGineH3irRgCRScCJAMoNLQVBpdIZdEQeI+6gFzJgRRp8TU0jgLHYDE4UB4kigIiF6nFUHkIo+aFZ6G1VDEM0gsDgX0q0M87i8IICwjRdQazTpHS6PT6UEGOn0djcWktJ1KAHp-aLgwA9AD8VBNTO0xMCtvNPuO-hEAHIEPgAO44PL4GT4TELGD4NAp3EEpUxgGk8mU3mfZym9HEQLK4isGCsbWBmnh4oUIA)):
```ts
import { z } from "zod";

const transformError = (error: Error) => error.message.toUpperCase();

const someSchema = z.object({
  username: z.string(),
});


const parseWithPrettifiedErrorMessage =
  <T extends z.ZodTypeAny>(schema: T) =>
  (content: string) => {
    const result = schema.safeParse(JSON.parse(content));

    if (result.success) {
      return result;
    }

    return {
      ...result,
      error: transformError(result.error),
    };
  };
  

export const parseSomeSchema = parseWithPrettifiedErrorMessage(someSchema);
//           ^? const parseSomeSchema: (content: string) => z.SafeParseSuccess<any> | { error: string; success: false; }

const result = parseSomeSchema('{"wrong": "content"}');

if (result.success) {
  const parsed = result.data;
  //    ^? const parsed: any
}
```

As you can see on the lines for `parseSomeSchema` and `parsed`, the return type in the case of an success is just `z.SafeParseSuccess<any>`, where I would like to have something more specific than `any`.

I tried to also use `z.ZodType` instead of `z.ZodTypeAny` but the only way I got this to work is to also explicitly specify the expected return value ([playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcBEyEAJvgNwBQFAxhAHYDO8MUAho5tCAKJQ5RwAvHAAUAUz7QAXHF78AlEIB8cCfwB0IMQwasA5mPUwIAVTBgJAYVYMxI+ZRr0mcBrjEBlagAsxIVkIo6hAARgBWYtQwIggUcHAArrZQdKxaMsjqTFDAdHr2ADQUaA5UMACeFnAe7l6+-oGZuZgSADwVFhCYrrU+fqxKjk6M8GCsULYA6sAw3gAKUGIwMMCYwGLEctAAstq6BkJxcK0A8gkwBXAAKqoAHjBidMQMQQBaJFeVYqfnKsKZ72InwsAEE6OUlCIGH1-DIropBEojiJaHQHmiZNlcnoESpYvF4qiXIsGAkADbwYTQ+qsLKsFpzca2EQAKQ8JwAcuoxhM7Kj0TB5KUjvFVqISeSYFkEtRqHtFPiCfFFjAEik4BKKZQlWgqEqVWq6IgRQT1GbNRcTfE1NI4Cx2AxOFAeJIoCILeobVB5EUddr0EMxLdILA4ETRkzPL0aYd4jypjN5iqVmsNlsoLsdPpvjUtHV+pC3HmYaxSgB6MtKqvxAB6AH4qOGNdpJYF41HizSRAByBD4ADuODy+Bk+H5jxg+DQ3eFYvdLYp0tl8uNhOcEd5xECHuIrBgrH9FaV9eKFCAA)):
```ts
import { z } from "zod";

const transformError = (error: Error) => error.message.toUpperCase();

const someSchema = z.object({
  username: z.string(),
});

type SomeSchema = z.infer<typeof someSchema>;


const parseWithPrettifiedErrorMessage =
  <Out, T extends z.ZodType<Out> = z.ZodTypeAny>(schema: T) =>
  (content: string) => {
    const result = schema.safeParse(JSON.parse(content));

    if (result.success) {
      return result;
    }

    return {
      ...result,
      error: transformError(result.error),
    };
  };


export const parseSomeSchema =
  parseWithPrettifiedErrorMessage<SomeSchema>(someSchema);
//           ^? const parseWithPrettifiedErrorMessage: <{ username: string; }, z.ZodTypeAny>(schema: z.ZodTypeAny) => (content: string)...

const result = parseSomeSchema('{"wrong": "content"}');

if (result.success) {
  const parsed = result.data;
  //    ^? const parsed: { username: string; }
}
```

As you can see, now `parseSomeSchema` and `parsed` are correctly typed. 

But: Is it also possible to let TypeScript infer the correct return type or do I always need to explicitly specify the result type?

**Comment by wujianguo** [comment]
```
async function myFetch<T, O>(url: string, schema: z.ZodType<T, z.ZodTypeDef, O>): Promise<z.infer<typeof schema>> {
  const resp = await fetch(url);
  const json = await resp.json();
  return schema.parse(json);
}

const GithubSchema = z.object({
  current_user_url: z.string(),
});
const result = await myFetch('https://api.github.com/', GithubSchema);
// const result: { current_user_url: string }
console.log(result);
```

**Comment by datenreisender** [comment, marked as answer]
Since asking my question, the documentation has been supplemented by the section [Inferring the inferred type](https://zod.dev/?id=inferring-the-inferred-type), which describes a solution. 

With it your `myFetch` can also be shortened to this:
```ts
async function myFetch<T extends z.ZodTypeAny>(url: string, schema: T) {
    const resp = await fetch(url);
    const json = await resp.json();
    return schema.parse(json) as z.infer<T>;
}
```

My original problematic code can be enhanced by changing the `return result` to `return result as z.SafeParseSuccess<z.infer<T>>`

Thanks for your input and reminding me of this question.

**Comment by markedwards** [comment]
This seems to work as far as TypeScript is concerned, but ESLint rejects this with:
```
Unsafe return of type `any` from function with return type `TypeOf<T>`.
```

The following example results in the ESLint error:
```
export function parsePayload<T extends z.ZodTypeAny>(
  schema: T,
  payload: unknown,
) {
  return schema.parse(payload) as z.infer<T>; // <- Unsafe return of type `any` from function with return type `TypeOf<T>`.
}
```

**Reply by markedwards** [reply]
It's possible to work around this by casting to `unknown`, but any time there's a `z.infer<T>` ESLint gets upset.  Can't this be typed in a way that doesn't require so much workaround?

**Reply by markedwards** [reply]
Maybe `z.ZodTypeAny` should in fact be `ZodType<unknown, ZodTypeDef, unknown>`?  That resolves the issue.

Or maybe there should be a `z.ZodTypeUnknown`?

**Reply by markedwards** [reply]
Created #4361 to follow this up.
